src/app/content/utils/seoUtils.ts

```diff
@@ -9,20 +12,167 @@ import {
+type PageTypes = 'page' | 'answer-key' | 'subpage' | 'eoc-page' | 'eob-page';
+
+interface DescriptionTemplateValues {
+  parentTitle: string;
+  parentType: 'chapter' | 'book' | 'other';
+  pageTitle: string;
+  bookTitle: string;
+  parentPrefix: string;
+}
+
+const domParser = new DOMParser();
+
+export const getParentPrefix = (node: LinkedArchiveTreeNode | undefined, includeTitle: boolean = false): string => {
   if (!node) {
     return '';
   }
 
   if (archiveTreeSectionIsChapter(node)) {
-    const number = getArchiveTreeSectionNumber(node);
-    return `Ch. ${number} `;
+    const number = getArchiveTreeSectionNumber(node).replace('Chapter', '').trim();
+    const name = getArchiveTreeSectionTitle(node);
+    return includeTitle ? `Ch. ${number}. ${name}` : `Ch. ${number} `;
   }
 
   return archiveTreeSectionIsBook(node.parent)
     ? getArchiveTreeSectionTitle(node) + ' '
-    : getParentPrefix(node.parent);
+    : getParentPrefix(node.parent, includeTitle);
+
+};
+
+const hideMath = (node: Element) => {
+  if (!node) {
+    return '';
+  }
+
+  node.querySelectorAll('.os-math-in-para').forEach((el: Element) => {
+    el.outerHTML = '...';
+  });
+  return node.textContent;
+};
+
+export const getTextContent = (str: string) => {
+  const parsed = domParser.parseFromString(str, 'text/html');
+  const text = parsed.body.textContent;
+  return text;
+};
+
+const removeExcludedContent = (node: HTMLElement) => {
+  if (!node) {
+    return null;
+  }
+  const excludedContent = node.querySelectorAll(
+    '[data-type="abstract"], .learning-objectives, .chapter-objectives, .be-prepared, .os-teacher'
+    ) || [];
+
+  for (let i = 0; i <= excludedContent.length; i++) {
+    if (excludedContent[i]) {
+      excludedContent[i].remove();
+    }
+  }
+};
+
+export const generateExcerpt = (str: string) => {
+  return str.replace(/\n+/g, ' ').replace(/\s+/g, ' ').trim().substring(0, 152) + '...';
+};
+
+const getPageType = (node: HTMLElement, values: DescriptionTemplateValues): PageTypes => {
+  const {parentType, parentPrefix, pageTitle} = values;
+  const nodeClasses = node.classList;
+  const nodeType = node.getAttribute('data-type');
+
+  if (nodeType === 'page' && !nodeClasses.contains('appendix')) {
+    return 'page';
+  } else if (pageTitle === parentPrefix) {
+    return 'eob-page';
+  } else if (
+    nodeClasses.contains('os-solution-container')
+      || nodeClasses.contains('os-solutions-container')
+    ) {
+    return 'answer-key';
+  } else if (parentType !== 'chapter' && parentType !== 'book') {
+    return 'subpage';
+  } else if (pageTitle !== parentPrefix) {
+    return 'eoc-page';
+  } else {
+    throw new Error('no page type detected');
+  }
+};
+
+const getPageDescriptionFromContent = (node: HTMLElement): string | null => {
+  if (!node) {
+    return null;
+  }
+  removeExcludedContent(node);
+  const paragraphs = node.querySelectorAll(':scope>section>p,:scope>p');
```

Comment:
@TomWoodward belated question about this: didn't we agree to prioritize paragraphs within sections over other paragraphs? This won't do that.

(It also throws an error because `:scope` yields an ID selector beginning with a digit, which [isn't allowed in CSS3](https://stackoverflow.com/a/37271406).)

Reply:
what element has a numeric id?

Reply:
the content div: https://share.getcloudapp.com/z8u6rmlY

Reply:
this seems ok unless we're concerned about paragraphs that are children of some other element: `const paragraphs = node.querySelectorAll('section>p') || node.querySelectorAll('p');`

Reply:
you're right that we need to prioritize paragraphs in sections, in order to simplify your filtering logic you might want to initialize as `[...node.querySelectorAll('section>p'), ...node.querySelectorAll('p')]` so that you can then just filter by text length

if you see anything with an invalid id that is a bug we should bring up with the content team. in your screenshot i think its valid because it starts with a `d`, but it looks like they're just using a uuid and _sometimes_ that will start with a number.